test/metabase/api/action_test.clj

```diff
@@ -114,6 +132,29 @@
           (let [request-body (mt/mbql-query categories {:filter [:= $id 22]})]
             (mt/user-http-request :crowberto :post 400 "action/row/delete" request-body))))))
 
+(defn- mock-requests
+  "Mock requests for testing validation for various actions. Don't use these for happy path tests! It's way too hard to
+  wrap your head around them. Use them for validating preconditions and stuff like that."
+  []
+  [{:action       "action/row/create"
+    :request-body (assoc (mt/mbql-query categories) :create-row {(format-field-name :name) "created_row"})
+    :expect-fn    (fn [result]
+                    ;; check that we return the entire row
+                    (is (schema= {:created-row {(format-field-name :id)   su/IntGreaterThanZero
+                                                (format-field-name :name) su/NonBlankString}}
+                                 result)))}
+   {:action       "action/row/update"
+    :request-body (assoc (mt/mbql-query categories {:filter [:= $id 1]})
+                         :update_row {(format-field-name :name) "updated_row"})
+    :expected     {:rows-updated [1]}}
+   {:action       "action/row/delete"
+    :request-body (mt/mbql-query categories {:filter [:= $id 1]})
+    :expected     {:rows-deleted [1]}}
+   {:action       "action/row/update"
+    :request-body (assoc (mt/mbql-query categories {:filter [:= $id 10]})
+                         :update_row {(format-field-name :name) "new-category-name"})
+    :expected     {:rows-updated [1]}}])
```

Comment:
This is untouched but I moved it down in the namespace a bit to try to make it clear it's not being used for the happy path `row/*` action tests anymore.